[riscv] Add support for RISC-V vector extension in arch v2

ita-sc opened the pull request.

ita-sc commented:
@DenisYaroshevskiy , @jfalcou created a new review as requested. Could you please take a look?

Please do note: removed changes related to `bit_cast`(for now). Going forward, I think that enabling bit_cast may faster, however this may be an improvement for the future.

jfalcou commented on include/eve/arch/riscv/as_register.hpp:
```diff
@@ -0,0 +1,299 @@
+}
+
+#  if defined(EVE_RETURN_M)
+#    error EVE_RETURN_M already defined
```
In which scenario can this happens ?

ita-sc replied:
This is just to be safe. I've already had an experience of debugging override of macros, so I prefer to have a guarantee that this situation is not happening here. If you wish, I can delete this, as now we do not have any possibility to have this macro defined.

jfalcou commented on include/eve/arch/riscv/as_register.hpp:
```diff
@@ -0,0 +1,299 @@
+    while( 0 )
+
+#  define EVE_RETURN_M(REG, LMUL)                                                                  \
+    do {                                                                                           \
```
the macro is used in a if consetxpr context, why not putting the { } in the if constexpr and just have the macro be the using + return. The do while(0) here is a bit overkill

DenisYaroshevskiy replied:
It was my suggetion - I liked the semicolon. If Joel doesn't like it - do the brackets

ita-sc replied:
Done

ita-sc commented:
> Do you think we can close the older prs for now? at leats the previous version of arch one?

Closed previous version of arch, full version for now moved to draft.

ita-sc commented:
Hi folks - do you have any opinion on how close we are to wrap up this review? To me it looks like we addressed most concerns - I just wanted to double-check if there are major issues that were not discussed yet

jfalcou commented:
> Hi folks - do you have any opinion on how close we are to wrap up this review? To me it looks like we addressed most concerns - I just wanted to double-check if there are major issues that were not discussed yet

I think this is good. @DenisYaroshevskiy ?

DenisYaroshevskiy commented:
I think this is close but every time I work through it again, I see small but important things to be fixed. 

I appreciate that it's frustrating but I wouldn't stop smth if I didn't think it was important.

@jfalcou - can you confirm that this tests aggregates?

DenisYaroshevskiy commented:
One iption is: we merge more or less as is and then I fix up small thingies afterwards myself.

Does that sound reasonable?

jfalcou commented:
The tests pass on the RISCV CI. So I am all for merging then fixinf the details

DenisYaroshevskiy commented:
@ita-sc - basically - up to you. I can continue giving you feedback. Or I can just fix it up after we merge.

If you feel like continue contributing your work further, please split it up futrher things in small chunks (function or two)- we will do those quickly and well

ita-sc commented:
> @ita-sc - basically - up to you. I can continue giving you feedback. Or I can just fix it up after we merge.

As main idea for this patch is clean and discussed, I suggest to merge this now, and make a separate review for small fix-ups later.

I will start work to create one more review that will clean some of TODOs here (like, extend `simd_cast` to remove functions like `riscv_lmul_trunc`) and next MR with commits to support next part of unit tests. 

Unfortunately, splitting commit one more time for basic tests seems quite hard -  for tests `top_bits` and `new_arch_first_tests` I almost sure that I will need full commit.
